### PR TITLE
meson: New build option to enable/disable tests

### PR DESCRIPTION
**Pull request by philn**

### CODE REVIEW

**Comment by philn** [comment]
Ping?

File: .travis.yml [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
 
 compiler:
   - g++
   - clang
 
 before_install:
   - sudo apt-get update -qq
-  - sudo apt-get install -qq nasm g++-multilib gcc-multilib libc6-dev-i386 python3-pip unzip
-  - sudo python3 -m pip install meson==0.44.1
+  - sudo apt-get install -qq nasm g++-multilib gcc-multilib libc6-dev-i386 python3-pip python3-setuptools unzip
```
Comment (MathieuDuponchelle) [inline .travis.yml]: why is this needed ooc?

File: test/meson.build [hunk 2]
```diff
@@ -1,15 +1,17 @@
-gtest_main_dep = dependency('gtest', main : true, fallback: ['gtest', 'gtest_main_dep'], required: false)
-gtest_dep = dependency('gtest', fallback: ['gtest', 'gtest_dep'], required: false)
+if not get_option('tests').disabled()
+  gtest_main_dep = dependency('gtest', main : true, fallback: ['gtest', 'gtest_main_dep'], required: false)
```
Comment (MathieuDuponchelle) [inline test/meson.build]: shouldn't `required` be set to true if `tests` was explicitly enabled?

File: .travis.yml (hunk 1 above)
Comment (philn) [reply .travis.yml]: IIRC there was a CI failure without this. I'll check again

**Comment by MathieuDuponchelle** [comment]
lgtm otherwise :)

File: .travis.yml (hunk 1 above)
Comment (philn) [reply .travis.yml]:
```
Collecting meson==0.47.0
  Downloading https://files.pythonhosted.org/packages/c0/9b/44cdb8adcbb186be6cba5c93718d0c68f177b0e8082ae00cafa63a1d3535/meson-0.47.0.tar.gz (1.2MB)
    100% |████████████████████████████████| 1.2MB 1.1MB/s 
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
    ImportError: No module named 'setuptools'
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-dhn678co/meson/
You are using pip version 8.1.1, however version 19.3.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
The command "sudo python3 -m pip install meson==0.47.0" failed and exited with 1 during .
```

**Comment by philn** [comment]
@FlyGoat @yinshiyou
please help me review the patch, thx.

**Comment by FlyGoat** [comment]
LGTM, For MIPS Part.

File: .travis.yml (hunk 1 above)
Comment (MathieuDuponchelle) [reply .travis.yml]: oic, ok :)

File: test/meson.build [hunk 3]
```diff
@@ -1,15 +1,17 @@
...
-  subdir('processing')
-endif
+  if gtest_dep.found()
+    subdir('api')
+    subdir('common')
+    subdir('decoder')
+    subdir('encoder')
+    subdir('processing')
+  endif
+endif
```
Comment (GuangweiWang) [inline test/meson.build]: please use the linux file format. this line break does not seem to be a linux-formatted line break.

File: .travis.yml [hunk 4]
```diff
@@ -1,14 +1,14 @@
 language: cpp
-dist: trusty
+dist: xenial
```
Comment (GuangweiWang) [inline .travis.yml]: why did you change the trusty to xenial?

Comment (philn) [reply .travis.yml]: As explained in the commit, Meson 0.47 requires Python 3.5.2 which is not shipped in Trusty.

File: test/meson.build (hunk 3 above)
Comment (philn) [reply test/meson.build]: This PR was prepared on a Linux machine. :)

**Comment by philn** [comment]
Ping?

File: .travis.yml (hunk 4 above)
Comment (GuangweiWang) [reply .travis.yml]: why meson must upgrade to 0.47?

Comment (philn) [reply .travis.yml]: The `feature` option type was added in Meson 0.47, as implied in the commit message.

**Comment by GuangweiWang** [comment]
LGTM. resolve the conflicts please. thanks.
